docs - updates for changes to resgroup MAX_SPILL_RATIO/MEMORY_LIMIT

lisakowen opened the pull request:
document the "fallback to statement_mem" and "do not reserve memory for RGs, use global shared memory" resource group changes.

i believe that the functionality is the same in master and 6X_STABLE (please confirm MPP team).  if so, this will be backported to 6X_STABLE.  i will submit a new PR for 5X_STABLE - will be similar, but without the catalog changes.

in this PR:
- "using" page - new MEMORY_LIMIT range and default, new MEMORY_SHARED_QUOTA default, new MEMORY_SPILL_RATIO default and new meaning to MEMORY_SPILL_RATIO=0, update examples, identify new default_group and admin_group limit defaults, update "low memory queries section to also set statement_mem"
- ALTER/CREATE RESOURCE GROUP - document the limit changes
- add statement_mem to list of GUCs in resource group category
- gp_toolkit - remove *proposed* attributes and related fields, update memory output
- gp_resgroup_config* views - remove some fields
- pg_resgroupcapability - remove proposed field
- memory_spill_ratio GUC - if 0, greenplum uses statement_mem
- best practices for RGs - misc edits, include memory_spill_ratio and statement_mem in list of config params, update the "low memory queries section

QUESTION:  it wasn't clear to me how to modify the statement_mem GUC description (http://docs-lisa-review.cfapps.io/6-0/ref_guide/config_params/guc-list.html#statement_mem)- right now, it is very specific to resource queues and utilizes gp_vmem_protect_limit in the calculation.

doc review site links:
-  "using" page:  http://docs-lisa-review.cfapps.io/6-0/admin_guide/workload_mgmt_resgroups.html#topic8339717
- ALTER RESOURCE GROUP:  http://docs-lisa-review.cfapps.io/6-0/ref_guide/sql_commands/ALTER_RESOURCE_GROUP.html#topic1
- best practices:  http://docs-lisa-review.cfapps.io/6-0/best_practices/resgroups.html
- gp_toolkit:  http://docs-lisa-review.cfapps.io/6-0/ref_guide/gp_toolkit.html#topic26x

dyozie commented on gpdb-doc/dita/admin_guide/workload_mgmt_resgroups.xml:
```diff
@@ -380,28 +380,34 @@
       <title>Additional Memory Limits for Role-based Resource Groups</title>
       <body>
-        <p>The memory reserved by a resource group for roles is further divided into fixed and
+        <p>If memory is reserved for a resource group for roles (non-zero
```
Maybe edit this to say "If resource group memory is reserved for roles (a non-zero MEMORY_LIMIT) ..." ?

dyozie commented on gpdb-doc/dita/admin_guide/workload_mgmt_resgroups.xml:
```diff
@@ -466,20 +472,26 @@ rg_perseg_mem = ((RAM * (vm.overcommit_ratio / 100) + SWAP) * gp_resource_group_
+           is 0.</p>
+          <p>When <codeph>MEMORY_SPILL_RATIO</codeph> is 0, Greenplum Database uses the
+            <xref href="../ref_guide/config_params/guc-list.xml#statement_mem" format="dita"><codeph>statement_mem</codeph></xref>
+            server configuration parameter value to control operator memory.</p>
```
operator memory -> query operator memory

zhangh43 replied:
Could also mention using `statement_mem` is used to make resource group compatible with resource queue.

zhangh43 replied:
Better query operator memory -> initial query operator memory.

p.s. actually we could not restrict the memory usage of an operator during runtime, this is just an initial value(expected but not guaranteed)

dyozie commented on gpdb-doc/dita/ref_guide/config_params/guc-list.xml:
```diff
@@ -6880,8 +6880,10 @@
+      <p>You can specify an integer percentage value from 0 to 100 inclusive. If you
+        specify a value of 0, Greenplum Database uses the
+            <xref href="#statement_mem" format="dita"><codeph>statement_mem</codeph></xref>
+            server configuration parameter value to control operator memory.</p>
```
operator memory -> query operator memory

zhangh43 replied:
query operator memory -> initial query operator memory

dyozie commented on gpdb-doc/dita/admin_guide/workload_mgmt_resgroups.xml:
```diff
@@ -380,28 +380,34 @@
         when you create the resource group. The minimum <codeph>MEMORY_LIMIT</codeph> percentage you
-        can specify for a resource group is 1, the maximum is 100.</p>
+        can specify for a resource group is 0, the maximum is 100. When <codeph>MEMORY_LIMIT</codeph>
+        is 0, Greenplum Database reserves no memory for the resource group, and uses resource
```
and uses -> but uses 

(global change)

dyozie replied:
It seems to me that this should be the spot where we also describe why someone would do this, even if just as an example (the resource queue tie-in?)  Right now it describes the mechanics but not the why of it.  It seems like there should be _some_ reference to resource queues somewhere if that's what drove the feature.

zhangh43 replied:
+1 on the resource queue tie-in example.
Set <codeph>MEMORY_LIMIT</codeph> to 0 is used to make resource group more compatible with resource queue, which let you use resource group more smoothly. 
It has two folds meaning. On one hand, queries in this kind of resource group have no reserved memory and have to compete the memory in resource group global shared memory with queries in other groups. On the other hand, it will increase the size of global shared memory region, which could share more system memory among different resource groups with no priority. 
Note that in a resource queue compatible resource group, there is no guaranty to allocate memory successfully and has risk of OOM when there are too many concurrent queries consuming memory from global shared memory region at the same time.
Considering a case that one resource group is important and you want to reduce the OOM risk of queries in this group compare with other groups, now you could reserve some private memory for this group by setting <codeph>MEMORY_LIMIT</codeph> bigger than 0.  Queries in this group will firstly allocate memories in its local memory region and results in lower risk of OOM.
Note that reserving group private memory will reduce the size of global shared memory. It's a tradeoff between fully use the system memory and reduce the oom risk for queries in important resource group.

gaos1 commented:
@lisakowen Thanks for the doc. Here are the answers to your questions:
1) right. the functionalities are the same on master and 6X.
2) for statement_mem GUC,  the calculation in resource group mode can be similar as resource queue mode, roughly: 
```
rg_perseg_mem = ((RAM * (vm.overcommit_ratio / 100) + SWAP) * gp_resource_group_memory_limit) / num_active_primary_segments
statement_mem = rg_perseg_mem/ max_expected_concurrent_queries
```

gaos1 commented:
maybe add a migration note for resource queue users:  memory_limit and memory_spill_ratio new capabilities provide a compatible memory management feature as resource queue. If you prefer to keep resource queue memory allocation and management features, please set memory_limit = 0 and memory_spill_ratio=0 for your groups.

zhangh43 commented on gpdb-doc/dita/best_practices/resgroups.xml:
```diff
@@ -69,6 +69,22 @@
+          </p>
+          <p>Set <codeph>memory_spill_ratio</codeph> to increase or decrease the initial
+            amount of memory Greenplum Database allots to a query. When you set
+            <codeph>memory_spill_ratio</codeph> to 0, Greenplum Database uses the
```
Change `the initial amount of memory Greenplum Database allots to a query` TO `the query operator memory amount for a query`

zhangh43 replied:
When memory_spill_ratio is bigger than 0, it represents a percentage of group's memory_limit value allocated to query operator. It also need to divide into the concurrency number of a resource group. If concurrency is high, the actual operator maybe small even if you set  memory_spill_ratio to the maximum value 100. In this case, set memory_spill_ratio to 0 and use statement_mem to control the operator memory.

zhangh43 commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_RESOURCE_GROUP.xml:
```diff
@@ -67,9 +67,14 @@ MEMORY_SPILL_RATIO <varname>integer</varname></codeblock>
+            <codeph>MEMORY_SPILL_RATIO</codeph> must also be 0.</pd>
+          <pd>The sum of the
             <codeph>MEMORY_LIMIT</codeph>s of all resource groups defined in the
             Greenplum Database cluster must not exceed 100.</pd>
```
Also mention the global shared memory region = 100 - sum of MEMORY_LIMIT of all resource groups here.

zhangh43 commented on gpdb-doc/dita/ref_guide/sql_commands/ALTER_RESOURCE_GROUP.xml:
```diff
@@ -78,14 +83,16 @@ MEMORY_SPILL_RATIO <varname>integer</varname></codeblock>
+            <codeph>MEMORY_SPILL_RATIO</codeph> value is 0. When <codeph>MEMORY_SPILL_RATIO</codeph>
+            is 0, Greenplum Database uses the
+            <xref href="../config_params/guc-list.xml#statement_mem" format="dita"><codeph>statement_mem</codeph></xref>
+            server configuration parameter value to control operator memory.</pd>
```
control operator memory to  control query operator memory

zhangh43 commented on gpdb-doc/dita/ref_guide/sql_commands/CREATE_RESOURCE_GROUP.xml:
```diff
@@ -63,15 +63,28 @@ MEMORY_LIMIT=<varname>integer</varname>
+            <codeph>MEMORY_SPILL_RATIO</codeph> value is 0. When <codeph>MEMORY_SPILL_RATIO</codeph>
+            is 0, Greenplum Database uses the
+            <xref href="../config_params/guc-list.xml#statement_mem" format="dita"><codeph>statement_mem</codeph></xref>
+            server configuration parameter value to control operator memory.</pd>
```
control operator memory to control query operator memory

zhangh43 commented on gpdb-doc/dita/best_practices/resgroups.xml:
```diff
@@ -69,6 +69,22 @@
+          <p>Set <codeph>memory_spill_ratio</codeph> to increase or decrease the initial
+            amount of memory Greenplum Database allots to a query. When you set
+            <codeph>memory_spill_ratio</codeph> to 0, Greenplum Database uses the
+            <codeph>statement_mem</codeph> setting to determine the query operator memory
```
query operator memory -> initial query operator memory

lisakowen commented:
i think that i incorporated the requested edits.  i added a resource queue mention to a new topic "About Reserving Resource Group Memory vs. Using Resource Group Global Shared Memory"  (not sure about the name) that discusses using reserved vs. global shared memory.

info about migrating from queues to groups will be addressed in a different PR - there are more considerations than just memory settings.

zhangh43 commented:
> i think that i incorporated the requested edits. i added a resource queue mention to a new topic "About Reserving Resource Group Memory vs. Using Resource Group Global Shared Memory" - http://docs-lisa-review.cfapps.io/6-0/admin_guide/workload_mgmt_resgroups.html#topic833fvs - (not sure about the name) that discusses using reserved vs. global shared memory. i updated the doc review site, you may need to refresh the page.
> 
> info about migrating from queues to groups will be addressed in a different PR - there are more considerations than just memory settings.

A new topic makes the explanation more clear, Thanks.
A few comments:
```
Any query submitted in the resource group competes for resource group global shared memory on a first-come, first-served basis with queries running in other groups.
```
In fact, running queries will request memory again and again during their lifecycles, So the query comes in first may be cancelled during its running due to OOM as well. 

```
About Reserving Resource Group Memory vs. Using Resource Group Global Shared Memory
```
As for name, I prefer About Using Reserved Resource Group Memory vs. Using Resource Group Global Shared Memory @pivotal-ning-yu what's your opinion?

lisakowen commented:
would like to merge this soon.  any more comments @zhangh43  or @dyozie ?

zhangh43 commented:
@lisakowen Looks good to me
